lib/metadata/acl.js

```diff
@@ -15,12 +15,14 @@ const acl = {
     addObjectACL(bucket, objectKey, objectMD, addACLParams, params, log, cb) {
         log.trace('updating object acl in metadata');
         function _didNotChange(grant) {
+            if (grant === 'Canned') {
+                return objectMD.acl.Canned === addACLParams.Canned;
+            }
             return objectMD.acl[grant].length === addACLParams[grant].length
                 && objectMD.acl[grant].every(value => addACLParams[grant].includes(value));
         }
-        const cannedChanged = objectMD.acl.Canned !== addACLParams.Canned;
-        if (cannedChanged || !_didNotChange('FULL_CONTROL') || !_didNotChange('WRITE_ACP')
-            || !_didNotChange('READ') || !_didNotChange('READ_ACP')) {
+        const isAclUnchanged = Object.keys(objectMD.acl).every(grant => _didNotChange(grant));
```

Comment:
that may not be correct: e.g. if `addACLParams` contains grants which are not present in `objectMD.acl`...
(until we have typescript to prove otherwise)

Reply:
When parsing the params from headers the function only considers the supported grants https://github.com/scality/cloudserver/blob/fe923f679decd55ffd0db7455910cc33adc0c2c1/lib/api/objectPutACL.js#L101-L109

Otherwise when the config is in a json format the function skips unsupported grants https://github.com/scality/cloudserver/blob/fe923f679decd55ffd0db7455910cc33adc0c2c1/lib/api/objectPutACL.js#L187-L189

And `objectMD.acl` is always initialised, and can never be removed, we always have all the grant fields even when empty

Reply:
This function is not objectPutACL, but in a separate module : so we should not make expectations on the calling code.
The only expectation is on the "contract" of the module (its interface) : e.g. what it expects, ...

For a typed-language, this is easy ; obviously here with javascript this is all blurry... The code in https://github.com/scality/cloudserver/blob/fe923f679decd55ffd0db7455910cc33adc0c2c1/lib/api/objectPutACL.js#L93-L99 kind of introduces such a type, but this is not used in other places... and there is always a risk of mistake (e.g. typo, or "new" field added in one place only...)

This will be much safer once we move to typescript, but until them, better avoid making too many assumptions, esp. when it can be fixed easily: e.g. if you just do an extra check on the number of keys we should be safe in any case...

> And objectMD.acl is always initialised, and can never be removed, we always have all the grant fields even when empty

fair enough.